File: oop_design_dattern/Adapter Pattern.py
from abc import ABC, abstractmethod


class JsonLogger(ABC):
    @abstractmethod
    def log_message(self, message: str) -> None:
        pass


class XmlLogger:
    def log(self, xml_message: str) -> None:
        print(xml_message)


class LoggerAdapter(JsonLogger):
    def __init__(self, xml_logger: XmlLogger) -> None:
        self.xml_logger = xml_logger

    def log_message(self, message: str) -> None:
        self.xml_logger.log(message)


logger = LoggerAdapter(XmlLogger())
logger.log_message("<message>hello</message>")


# Practice
class Square:
    def __init__(self, sideLength=0.0):
        self.sideLength = sideLength

    def getSideLength(self) -> float:
        return self.sideLength


class SquareHole:
    def __init__(self, sideLength: float):
        self.sideLength = sideLength

    def canFit(self, square: Square):
        return self.sideLength >= square.getSideLength()


class Circle:
    def __init__(self, radius: float):
        self.radius = radius

    def getRadius(self):
        return self.radius


class CircleToSquareAdapter(Square):
    def __init__(self, circle: Circle):
        self.circle = circle

    def getSideLength(self) -> float:
        return 2 * self.circle.getRadius()

File: oop_design_pattern/Adapter Pattern.java
interface JsonLogger {
  void logMessage(String message);
}

class XmlLogger {
  public void log(String xmlMessage) {
    System.out.println(xmlMessage);
  }
}

class LoggerAdapter implements JsonLogger {
  private final XmlLogger xmlLogger;

  public LoggerAdapter(XmlLogger xmlLogger) {
    this.xmlLogger = xmlLogger;
  }

  @Override
  public void logMessage(String message) {
    xmlLogger.log(message);
  }
}

class Main {
  public static void main(String[] args) {
    JsonLogger logger = new LoggerAdapter(new XmlLogger());
    logger.logMessage("<message>hello</message>");
  }
}


// Practice
class SquareHole {
    private double sideLength;

    public SquareHole(double sideLength) {
        this.sideLength = sideLength;
    }

    public boolean canFit(Square square) {
        return this.sideLength >= square.getSideLength();
    }
}

class Square {
    private double sideLength;

    public Square() {}

    public Square(double sideLength) {
        this.sideLength = sideLength;
    }

    public double getSideLength() {
        return this.sideLength;
    }
}

class Circle {
    private double radius;

    public Circle(double radius) {
        this.radius = radius;
    }

    public double getRadius() {
        return this.radius;
    }
}

class CircleToSquareAdapter extends Square {
    private Circle circle;

    public CircleToSquareAdapter(Circle circle) {
        this.circle = circle;
    }

    @Override
    public double getSideLength() {
        return 2 * this.circle.getRadius();
    }
}
